domain/module.go

```diff
@@ -13,4 +16,10 @@ type Module interface {
 	Apply(r *Resource) (ResourceStatus, error)
 	Validate(r *Resource) error
 	Act(r *Resource, action string, params map[string]interface{}) (map[string]interface{}, error)
+	Log(ctx context.Context, r *Resource, filter map[string]string) (chan LogChunk, error)
```

Comment:
the return channel should be changed from `chan LogChunk` to `<-chan LogChunk` I think.

Reply:
Will make the change for the read-only channel.
<hr>

For the module I was thinking the module itself can also throw a not implemented error if needed, leaving the choice to the module creator. (It's being done for the firehose module as of now in his PR)
What would be advantage of using the optional interface approach?

Reply:
with optional interface approach, 
* the behaviour of `Log()` for modules that don't support it will be defined at the `module-service` level making it consistent (since each module isn't responsible for returning whatever error it wants when log is not supported).. 
* module implementor will not have to add a stub `Log()` method which only does `return nil, errNotSupported` just for the sake of satisfying interface.
* clarity in terms of what a module actually supports and what it doesn't just by looking at the methods it has -- for example, in godoc. i.e., when you see a module without `Log()` method it is clear it does not support log stream. without optional interface, you have to look into the module code to confirm it because every module will have that method. 

And, with the optional interface pattern, the modules are still free to have a Log() method and return an explicit error to indicate not-supported as well. so it is a superset of both approaches.

Reply:
Got it. Let me do that.